lib/src/rules/prefer_is_empty.dart

```diff
@@ -0,0 +1,163 @@
+import 'package:linter/src/linter.dart';
+
+const desc = 'Use isNotEmpty for Iterables and Maps.';
+
+const useIsNotEmpty = 'Use isNotEmpty instead of length';
+const useIsEmpty = 'Use isEmpty instead of length';
+const alwaysFalse = 'Always false because length is always greater or equal 0.';
+const alwaysTrue = 'Always true because length is always greater or equal 0.';
+
+const details = '''
+**DO NOT** use `.length` to see if a collection is empty.
+
+The `Iterable` contract does not require that a collection know its length or be
+able to provide it in constant time. Calling `.length` just to see if the
+collection contains anything can be painfully slow.
+
+Instead, there are faster and more readable getters: `.isEmpty` and
+`.isNotEmpty`. Use the one that doesn’t require you to negate the result.
+
+**GOOD:**
+```
+if (lunchBox.isEmpty) return 'so hungry...';
+if (words.isNotEmpty) return words.join(' ');
+```
+
+**BAD:**
+```
+if (lunchBox.length == 0) return 'so hungry...';
+if (words.length != 0) return words.join(' ');
+```
+''';
+
+class _LintCode extends LintCode {
+  static final registry = <String, LintCode>{};
+
+  factory _LintCode(String name, String message) => registry.putIfAbsent(
+      name + message, () => new _LintCode._(name, message));
+
+  _LintCode._(String name, String message) : super(name, message);
+}
+
+class PreferIsEmpty extends LintRule {
+  PreferIsEmpty()
+      : super(
+            name: 'prefer_is_empty',
+            description: desc,
+            details: details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+
+  void reportLintWithDescription(AstNode node, String description) {
+    if (node != null) {
+      reporter.reportErrorForNode(new _LintCode(name, description), node, []);
+    }
+  }
+}
+
+class Visitor extends SimpleAstVisitor {
+  final PreferIsEmpty rule;
+  Visitor(this.rule);
+
+  @override
+  visitSimpleIdentifier(SimpleIdentifier identifier) {
+    AstNode lengthAccess;
+    SimpleIdentifier lengthIdentifier;
+    InterfaceType type;
+
+    AstNode parent = identifier.parent;
+    if (parent is PropertyAccess) {
+      lengthIdentifier = parent.propertyName;
+      lengthAccess = parent;
+      if (parent.target?.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.target?.bestType;
+    } else if (parent is PrefixedIdentifier) {
+      lengthIdentifier = parent.identifier;
+      lengthAccess = parent;
+      if (parent.prefix.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.prefix.bestType;
+    }
+
+    if (lengthIdentifier == null) {
+      return;
+    }
+
+    // Should be "length".
+    Element propertyElement = lengthIdentifier.bestElement;
+    if (propertyElement?.name != 'length') {
+      return;
+    }
+
+    // Should have `isNotEmpty` and `isEmpty`.
+    if (type.getGetter('isEmpty') == null ||
+        type.getGetter('isNotEmpty') == null) {
+      return;
```

Comment:
This doesn't restrict the lint to invocations on subtypes of `Iterable`. Consider:
```
class BankAccount {
  /// Return true if the balance is zero or negative.
  bool get isEmpty => ...;
  /// Return true if the balance is positive.
  bool get isNotEmpty => ...;
  /// Return the length of time (in days) that this account has been open.
  Duration get length => ...;
}
```
You might want to test whether the `type.isSubtypeOf(typeProvider.iterable)`.

If not, then note that `getGetter` returns the getter defined in the class, but not one that is inherited. So, a class like
```
class MyList extends List {
  void rotateContents() {}
}
```
will return `null` for both getters. If you really want to use duck typing semantics, then you'll need to use `lookUpGetter`.